src/optix_coop_vec.cpp

```diff
@@ -0,0 +1,488 @@
+        default: jitc_fail("get_coop_vec_type_id(): unsupported operation!");
+    }
+}
+
+uint32_t jitc_optix_coop_vec_type_id(VarType vt) {
+    switch (vt) {
+        case VarType::Float16: return 0x2A01;
+        case VarType::Float32: return 0x2A03;
+        case VarType::UInt32: return 0x2A08;
+        case VarType::Int32: return 0x2A09;
+        default:
+            jitc_fail("jitc_optix_coop_vec_type_id(): unsupported variable type!");
+    }
+}
+
+uint32_t jitc_optix_coop_vec_layout_id(MatrixLayout ml) {
+    switch (ml) {
+        case MatrixLayout::TrainingOptimal: return OPTIX_COOP_VEC_MATRIX_LAYOUT_TRAINING_OPTIMAL;
+        case MatrixLayout::InferencingOptimal: return OPTIX_COOP_VEC_MATRIX_LAYOUT_INFERENCING_OPTIMAL;
+        case MatrixLayout::RowMajor: return OPTIX_COOP_VEC_MATRIX_LAYOUT_ROW_MAJOR;
+        default:
+            jitc_fail("jitc_optix_coop_vec_layout_id(): unsupported layout type!");
+    }
+}
+
+// --------------------------------------------------------------------------
+// Helper routines to generate frequently used PTX fragments
+
+static uint32_t get_reg_count(uint32_t l) {
+    if (l <= 16)
+        return 16;
+    else if (l <= 64)
+        return 64;
+    else
+        return 0;
+}
+
+static void declare_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+
+    fmt("        .local .align $u .b8 $s[$u];\n"
+        "        .reg.u64 %$s;\n"
+        "        cvta.local.u64 %$s, $s;\n",
+        tsize, name, tsize * length,
+        name,
+        name, name);
+}
+
+static void copy_to_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+    for (uint32_t i = 0; i < length; ++i)
+        fmt("        st.local.$b [$s+$u], %cv$u_$u;\n", v, name, tsize * i, v->reg_index, i);
+}
+
+static void copy_from_buffer(const char *name, const Variable *v) {
+    uint32_t tsize  = type_size[v->type],
+             length = v->array_length;
+
+    for (uint32_t i = 0; i < length; ++i)
+        fmt("        ld.local.$b %cv$u_$u, [$s+$u];\n", v, v->reg_index, i, name, tsize * i);
+}
+
+static void put_elem(const Variable *v, uint32_t reg_count, bool trailing_comma, bool is_return_value = false) {
+    for (uint32_t i = 0; i < reg_count; ++i) {
+        if (i < v->array_length || is_return_value)
+            fmt("%cv$u_$u, ", v->reg_index, i);
+        else
+            put("%u, ");
+    }
+    if (!trailing_comma)
+        buffer.rewind(2);
+}
+
+void jitc_optix_render_coop_vec_unpack(const Variable *v, const Variable *a0) {
+    uint32_t tsize = type_size[v->type];
+
+    put("    // coop_vec_unpack\n");
+    if (tsize == 4) {
+        fmt("    mov.b32 $v, %cv$u_$u;\n", v, a0->reg_index, (uint32_t) v->literal);
+    } else {
+        fmt("    {\n"
+            "        .reg.$b %temp;\n"
+            "        cvt.u$u.u32 %temp, %cv$u_$u;\n"
```

Comment:
In theory we could back e.g. two `f16` entries of a cooperative vector into one of these `.u32` registers, but I imagine there would be hardly any gains and it wouldn't be worth the complexity?
Just checking because registers are at a premium.

Reply:
The backend (OptiX) may or may not perform such an optimization for us. If you look at how user code interfaces with OptiX through the builtin intrinsics, this is the way to do it officially. None of the OptiX SDK examples use the `float16x2` packed arithmetic types.